gcn_classic_to_json/utils.py

```diff
@@ -0,0 +1,55 @@
+import numpy as np
+from astropy.time import Time
+
+
+def datetime_to_iso8601(date, time):
+    """Converts time to ISO 8601 format.
+
+    The function converts input into the ISO 8601 format from
+    Truncated Julian Date by first converting it to Julian Date.
+
+    Parameters
+    ----------
+    date : int
+        Date must be in Truncated Julian Date format.
+    time : int
+        Time of day must in Seconds of Day format.
+
+    Returns
+    -------
+    string
+        returns datetime in ISO8601 format.
+
+    Notes
+    -----
+    The zero point for Truncated Julian Day is given in https://en.wikipedia.org/wiki/Julian_day.
+    """
+    TJD0 = (2440000, 0.5)
+    return Time(date + TJD0[0], time / 8640000 + TJD0[1], format="jd").isot + "Z"
+
+
+def binary_to_string(binary):
+    """Converts a binary array to a ASCII-string.
+
+    The function converts each field encoded as a 4-byte integer into
+    four 1-byte integers and then to their corresponding ASCII value.
+
+    Parameters
+    ----------
+    binary : array-like
+        A array for binary values encoded as 4-byte integers.
+
+    Returns
+    -------
+    string:
+        returns the corresponding ASCII string.
+
+    Notes:
+    ------
+    The strings in the binary packets seem to encoded little-endian.
+    """
+    bits_array = np.asarray(binary, dtype="<i4").view(dtype="u1")
+    str_list = [chr(bits) for bits in bits_array]
+    string = "".join(str_list)
+    excess_char = string.count("\u0000")
+    return string.replace("\u0000", "", excess_char)
```

Comment:
```suggestion
    return binary.tobytes().strip(b'\0').decode()
```

Reply:
This change gives url:  "http://cgbm.calet.jp/cgbm_trigger/flight/ednith.x\u0000\u0000lm"
as opposed to: "http://cgbm.calet.jp/cgbm_trigger/flight/index.html"

Reply:
What. The. F***. You're right. These strings are written as 4-byte stings, with each 4 bytes written backwards. I have never seen such a weird encoding.

```
$ od -t c gcn_classic_to_json/notices/CALET_GBM_FLT_LC/example.bin 
0000000   \0  \0  \0 240  \0  \0  \0   )  \0  \0  \0  \0  \0   K 377 334
0000020    S   n   , 352  \0  \0   O 334  \0   K 366   V  \0 035 224   D
0000040   \0  \a 257   &  \0  \0 003   >  \0  \0 001 220  \0  \0 006   @
0000060   \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0
0000100  023 236   ,   {  \0 346  \0   (  \0  \0  \0  \0  \0  \0  \0 004
0000120   \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0
*
0000160   \0  \0  \0  \0   e   d   n   i   t   h   .   x  \0  \0   l   m
0000200   \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0
0000220   \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \0  \n
0000240
```

Reply:
```suggestion
    return np.fliplr(data.view('c').reshape(-1, 4)).ravel().tobytes().strip(b'\0').decode()
```

Reply:
There seems to be a problem with binary.view('c'), it returns the following:
ValueError: Circular reference detected

Reply:
This should be:

```suggestion
    return np.fliplr(binary.view('c').reshape(-1, 4)).ravel().tobytes().strip(b'\0').decode()
```

and I do get an exception if I apply that patch, but it's a different exception:

```
$ poetry run pytest
=================================================== test session starts ====================================================
platform darwin -- Python 3.12.4, pytest-8.3.1, pluggy-1.5.0
rootdir: /Users/lpsinger/src/gcn-classic-to-json
configfile: pyproject.toml
plugins: cov-5.0.0, gcn-classic-to-json-0.1.0
collected 4 items                                                                                                          

gcn_classic_to_json/test/test_notices.py Fx..                                                                        [100%]

========================================================= FAILURES =========================================================
__________________________________________ test_all_fields_used[CALET_GBM_FLT_LC] __________________________________________

key = 'CALET_GBM_FLT_LC', monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x102627470>

    @pytest.mark.parametrize(
        "key",
        keys_passing_except_for("SWIFT_BAT_GRB_POS_ACK"),
    )
    def test_all_fields_used(key, monkeypatch):
        """Check that every field in the binary packet is used in the conversion."""
        used = np.zeros(40, dtype=bool)
    
        class NDArrayNanny(np.ndarray):
            def __getitem__(self, i):
                used[i] = True
                return super().__getitem__(i)
    
        def mock_frombuffer(*args, **kwargs):
            return _orig_frombuffer(*args, **kwargs).view(NDArrayNanny)
    
        monkeypatch.setattr(notices, "_frombuffer", mock_frombuffer)
    
        bin_path = files / key / "example.bin"
        value = bin_path.read_bytes()
>       notices.parse(key, value)

gcn_classic_to_json/test/test_notices.py:40: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
gcn_classic_to_json/notices/__init__.py:22: in parse
    return parser(ints)
gcn_classic_to_json/notices/CALET_GBM_FLT_LC/__init__.py:51: in parse
    + utils.binary_to_string(bin[29:39]),
gcn_classic_to_json/utils.py:51: in binary_to_string
    return np.fliplr(binary.view('c').reshape(-1, 4)).ravel().tobytes().strip(b'\0').decode()
../../Library/Caches/pypoetry/virtualenvs/gcn-classic-to-json-J-3iCvkw-py3.12/lib/python3.12/site-packages/numpy/lib/twodim_base.py:99: in fliplr
    return m[:, ::-1]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[IndexError('too many indices for array: array is 1-dimensional, but 2 were indexed') raised in repr()] NDArrayNanny object at 0x12468f250>
i = (slice(None, None, None), slice(None, None, -1))

    def __getitem__(self, i):
>       used[i] = True
E       IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed

gcn_classic_to_json/test/test_notices.py:30: IndexError
================================================= short test summary info ==================================================
FAILED gcn_classic_to_json/test/test_notices.py::test_all_fields_used[CALET_GBM_FLT_LC] - IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed
========================================== 1 failed, 2 passed, 1 xfailed in 0.19s ==========================================
```

But this is a problem in the monkeypatch, not in the code itself.

Reply:
That will be fixed by https://github.com/nasa-gcn/gcn-classic-to-json/pull/33.

Reply:
Fixed now, please rebase.